### ISSUE TITLE
data wrangler issue for debugging in python

### ISSUE THREAD

**Issue by m-fariz-a** [posit-dev/positron#5659]
Positron versiion: 2024.12.0 build 96
OS version: Ubuntu 24-04

## Issue:

When I debug a python file using `python debugger` built in on the right-top corner button and reach a break point, `View Value in Data Viewer` still expecting `Data Wrangler` extension from Microsoft. 

I hope this is already in progress.

## Expected or desired behavior:

* It open the `Data Explorer`

**Comment by juliasilge** [comment]
We currently _bundle_ the Python debugger and have not forked it (i.e. have not made any changes to that extension):
https://github.com/posit-dev/positron/blob/c7ef34fc7078ac7ecf13ff3328605dfba93f3eea/product.json#L194

We would have to look into how it's possible for us to hook into what's going on in the debugger.

**Comment by isabelizimm** [comment]
This is actually coming from [the jupyter extension](https://github.com/microsoft/vscode-jupyter/blob/9f67a495511328861b8e6196c5b4b498e317f4c0/src/webviews/extension-side/dataviewer/dataViewerDelegator.ts#L15). If possible, we will want to register the Data Explorer as an "external data viewer" for Jupyter to open up.

**Comment by isabelizimm** [comment]
Notes from exploration:

- There is a Command that registers this support: [`jupyter.showDataViewer`](https://github.com/microsoft/vscode-jupyter/blob/245979fbb9ae3c27eefef9777c16b8da5dd00d04/package.json#L673)
- Another relevant Command is the [`jupyter.showJupyterDataViewer`](https://github.com/microsoft/vscode-jupyter/blob/245979fbb9ae3c27eefef9777c16b8da5dd00d04/package.json#L2038-L2054) where the data viewer is created.
- We are not able to ignore these commands as [we do with interactive window support](https://github.com/posit-dev/positron/blob/559655deef4e4722d7cc982b66eae4d872081a69/src/vs/workbench/api/common/configurationExtensionPoint.ts#L270-L277) since it is not a _configuration_ in the `package.json`, rather, it is a Command.
- We can add in the Positron Data Explorer by adding a `jupyterVariableViewer` into the `positron-python/package.json`, see below for example. However, since the debugger runs in the terminal, not the console, the variables pane does not have the context to open up the data frame object. The `vscode-jupyter` extension uses a [function to convert debug protocol variables into data viewer-compatible variables](https://github.com/microsoft/vscode-jupyter/blob/245979fbb9ae3c27eefef9777c16b8da5dd00d04/src/kernels/variables/helpers.ts#L18-L32).

```
...
"contributes":  {
    "jupyterVariableViewers": [
      {
        "command": "python.showDataExplorer",
        "title": "Show Data Explorer",
        "dataTypes": [
          "DataFrame",
          "Series",
          "Tensor",
        ]
      }
    ],
}
...
```

**Comment by seeM** [comment]
## Steps to reproduce the issue:

1. Create a Python file that defines a dataframe e.g.
    
    ```python
    import pandas as pd
    df = pd.DataFrame()
    print(df)
    ```
2. Set a breakpoint on the line `print(df)`.
3. Run the _Python: Debug Python file in Terminal_ command, it should stop at the breakpoint.
4. In the **Run and Debug** pane, **Debug Variables** section, right click `df` and click **View Value in Data Viewer**.

You will see a popup:

<img width="260" height="241" alt="Image" src="https://github.com/user-attachments/assets/82235cb2-8760-41a3-83a1-9b151a233c2d" />

## Expected or desired behavior:

The title should be something like **View Data Table** (or **View Data Table in Data Explorer**?).

It should open the object in Positron's Data Explorer.

**Comment by seeM** [comment]
This should also be fixed when debugging notebooks. We can track that separately if needed.

**Comment by juliasilge** [comment]
Requested again in https://github.com/posit-dev/positron/discussions/8546#discussioncomment-13985693

I think you can't use a workaround like `%view dataframe` in Python, since you aren't debugging in the console, right?

**Comment by m-fariz-a** [comment]
Right now, python debugging works with `debug console` and you can see in the terminal like this

> /usr/bin/env /home/fariz/Python/bin/python /home/fariz/.positron/extensions/ms-python.debugpy-2025.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher 33061 -- /home/fariz/Documents/mypython.py

The user default debug config is like this

```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python Debugger: Python File",
            "type": "debugpy",
            "request": "launch",
            "program": "${file}"
        }
    ]
}
```

May be if there's a way so that python debugging session will use positron console instead of the default `debug console` it will work, but I'm not really sure whether it will clear the variables after the debug session end like in VSCode if using positron console.

**Comment by seeM** [comment]
@juliasilge I don't think you can use `%view` while debugging due to #4350.

**Comment by seeM** [comment]
Yes, you can't use a workaround like `%view dataframe` since it's debugging via the terminal which is totally external to Positron. There isn't actually a kernel, just a Python process which launches a debugpy server.

Getting this to work will actually require kernel-aware Python debugging, which I don't think we're tracking yet. That would probably use the same mechanism used for debugging notebooks (#3906), but will need additional work.

In addition to kernel-aware Python debugging, we'll also need to address #4350.

**Comment by cddesja** [comment]
Is there any update on this? I am an admin of a Posit Workbench environment and my users of Positron are wondering if/when it might be possible to use Positron's Data Viewer when in debug mode.

**Comment by juliasilge** [comment]
@cddesja We've got this prioritized to the backlog for this year, which is not a hard commitment to a specific monthly milestone but a plan to address this in the medium term.

**Comment by cddesja** [comment]
Thanks. I appreciate the update.
